### ISSUE TITLE
Provide a state store implementation for Couchbase

### ISSUE THREAD

**Issue by abhirockzz** [dapr/components-contrib#98]
## Describe the proposal

Provide a state store implementation for [Couchbase](https://docs.couchbase.com/home/index.html)

**Comment by abhirockzz** [comment]
I am currently looking into this.

**Comment by abhirockzz** [comment]
Implementation questions

### Which state spec to use?

Gitter conversation

```
Abhishek Gupta @abhirockzz Nov 03 11:34
Some observations reg "Expected state store behaviors" as per https://github.com/dapr/docs/blob/master/reference/api/state.md#expected-state-store-behaviors
- Currently this is not really enforced via an interface
- Implementation needs to map it to respective state store concepts (which if fine, but tricky at times)
- The implementation itself requires "if-else" for concurrency and consistency, and then choosing the right "API call"

Abhishek Gupta @abhirockzz Nov 03 11:40
I wonder if it's possible to make this easier? Also, I see that some of the state store implementations do not take this (consistency, concurrency etc.) into account

Yaron Schneider @yaron2 Nov 03 12:33
Yes, that's why the consistency/concurrency are optional as part of the request. A state store does not have to implement them. We will extend the docs on supported state stores to extend the table and show which stores enable those features.
That doc is outdated.. it references a spec that is not in use anymore. These behaviors are not required.

Also, state store operations section is missing from https://github.com/dapr/docs/blob/master/reference/api/state.md#state-operation-options. Raised https://github.com/dapr/docs/issues/225
```

### How to pass additional metadata?

Gitter conversation

```
Abhishek Gupta @abhirockzz Nov 03 22:53
Thanks. For Dapr state stores, consistency is either "eventual" or "strong", but that is not the case with every database. Classic e.g. is Cassandra which has the equivalent of ALL (all replicas) or ANY (at least one node) for "strong" and "eventual" respectively, but also has other consistency levels e.g. QUORUM, ONE, TWO etc. How can state store implementations account for these?

Yaron Schneider @yaron2 00:20
Because Dapr defines eventual or strong, a state store can either account for these two, or none. In order to account for extra settings, implementations can use the metadata fields that can be set with every state request, and map these to the specific implementation of the store.
```

### Is there a way to pass in additional metadata?

Fir e.g. I would like the bucket name to be passed in per operation rather than be configured at an app level (in Dapr config YAML).

**Comment by yaron2** [comment]
In order to pass in a custom property such as `bucket`, users can send an optional metadata map on every request. See [here](https://github.com/dapr/components-contrib/blob/4bbfb82e98fcf5a91846d3e7fca54e4457bc74b5/state/requests.go#L36).

So in the Couchbase implementation, you can pull that property from `metadata["bucket"]`.

I recommend providing a default bucket name, and a metadata property on the YAML itself in addition to allowing developers to override them with a per request metadata field.

**Comment by yaron2** [comment]
State management doc has been updated. Thanks for pointing out!

**Comment by abhirockzz** [comment]
> State management doc has been updated. Thanks for pointing out!

Great! But, as per Gitter, [this section](https://github.com/dapr/docs/blob/master/reference/api/state.md#optional-behaviors) is not up to date, correct?

**Comment by abhirockzz** [comment]
> In order to pass in a custom property such as `bucket`, users can send an optional metadata map on every request. See [here](https://github.com/dapr/components-contrib/blob/4bbfb82e98fcf5a91846d3e7fca54e4457bc74b5/state/requests.go#L36).
> 
> So in the Couchbase implementation, you can pull that property from `metadata["bucket"]`.
> 
> I recommend providing a default bucket name, and a metadata property on the YAML itself in addition to allowing developers to override them with a per request metadata field.

Thanks - having the default name with override option sounds good! I did the metadata option earlier, but raised this question since it's only for `Set` operation (unless I am missing something?)

**Comment by yaron2** [comment]
Yes, we'll need to add the metadata property to Get and Delete requests as well.

It's a small change to the struct, but I recommend you continue with the default bucket and yaml field impl and add the overridable buckets in a separate PR once we add support for Get and Delete.

**Comment by abhirockzz** [comment]
> Yes, we'll need to add the metadata property to Get and Delete requests as well.
> 
> It's a small change to the struct, but I recommend you continue with the default bucket and yaml field impl and add the overridable buckets in a separate PR once we add support for Get and Delete.

Sounds good. I guess I should wait on the updated "State management" spec before working on details related to consistency etc. ?

**Comment by yaron2** [comment]
Spec was updated.

**Comment by abhirockzz** [comment]
Additional notes on consistency implementation

For consistency tuning, Couchbase Go SDK provides [`ReplaceDura`](https://godoc.org/gopkg.in/couchbase/gocb.v1#Bucket.ReplaceDura)

```
func (b *Bucket) ReplaceDura(key string, value interface{}, cas Cas, expiry uint32, replicateTo, persistTo uint) (Cas, error)
```
It needs information about how many replicas are there ([see docs](https://docs.couchbase.com/go-sdk/current/durability.html#configuring-durability)), based on which `replicateTo` and `persistTo` can be passed to `ReplaceDura` func. Although no. of replicas can be added as Dapr metadata, but it feels "hacky". I probably need to check whether this can be obtained at runtime via SDK itself, but not sure if this is "good enough"

**Comment by yaron2** [comment]
> Additional notes on consistency implementation
> 
> For consistency tuning, Couchbase Go SDK provides [`ReplaceDura`](https://godoc.org/gopkg.in/couchbase/gocb.v1#Bucket.ReplaceDura)
> 
> ```
> func (b *Bucket) ReplaceDura(key string, value interface{}, cas Cas, expiry uint32, replicateTo, persistTo uint) (Cas, error)
> ```
> 
> It needs information about how many replicas are there ([see docs](https://docs.couchbase.com/go-sdk/current/durability.html#configuring-durability)), based on which `replicateTo` and `persistTo` can be passed to `ReplaceDura` func. Although no. of replicas can be added as Dapr metadata, but it feels "hacky". I probably need to check whether this can be obtained at runtime via SDK itself, but not sure if this is "good enough"

Did you get to check on this?

**Comment by abhirockzz** [comment]
I will look into this once I am back to work next week (second half)

**Comment by abhirockzz** [comment]
There is a [`GetBuckets` function] (https://godoc.org/gopkg.in/couchbase/gocb.v1#ClusterManager.GetBuckets) which does return the no. of replicas of a bucket. But it is at a cluster level and iterated through "all" the buckets

**Comment by yaron2** [comment]
So is it possible to get all buckets and then count the replicas for the specific bucket that was configured? what's your recommendation?

**Comment by abhirockzz** [comment]
Yes, that's what the function is for. But I was a little apprehensive about it, given the fact that this will be repeated for each state operation (which requires strong consistency)

**Comment by yaron2** [comment]
Yeah checking for every operation is not performant.

Two suggestions:

1. We add an optional `replicas` field to the metadata
2. We check for the number of replicas for the bucket on Init and use that value

For strong consistency we don't need to wait for all replicas, just for a quorum.

**Comment by abhirockzz** [comment]
Thanks. I will add the `replicas` field to the metadata and implement it that way
